File: BinarySearchTree/C/E783-minimum-distance-between-bst-nodes.c
// time:  O(n)
// space: O(n)

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

void treeToNums(struct TreeNode* root, int* nums, int* i){
    if(root->left)
        treeToNums(root->left, nums, i);
    
    nums[*i] = root->val;
    (*i)++;
    
    if(root->right)
        treeToNums(root->right, nums, i);
}
int minDiffInBST(struct TreeNode* root) {
    int* nums = (int*)malloc(100 * sizeof(int));
    int* i = (int*)malloc(sizeof(int));
    *i = 0;
    treeToNums(root, nums, i);
    
    int m = 65536;
    for(int j = 0; j < (*i)-1; j++){
        int t1 = nums[j+1] - nums[j];
        m = t1<m ? t1 : m;
    }
    return m;
}

File: BinarySearchTree/Python3/E783-minimum-distance-between-bst-nodes.py
# time:  O(n)
# space: O(h)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def minDiffInBST(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def TtN(n, l):
            if n.left:
                TtN(n.left, l)
                
            l.append(n.val)
            
            if n.right:
                TtN(n.right, l)
                
        l = []
        TtN(root, l)
        return min([abs(a - b) for a, b in zip(l, l[1:])])